### DISCUSSION TITLE
How to Put All Your Code in the Appendix in R Markdown | /en/2018/09/code-appendix/

### DISCUSSION THREAD

**Discussion by yihui** [rbind/yihui#1206] [Comments]
https://yihui.org/en/2018/09/code-appendix/

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Alex was talking about generating PDFs for his homework assignments, so code folding won't work. If the output format can be HTML, you are certainly right that folding the code chunks will be nicer.

> Originally posted on 2018-09-19 02:47:14

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Glad to know the post was useful to you! And of course, thank you for sharing the original trick, without which I wouldn't start writing this post.

> Originally posted on 2018-09-19 21:51:21

**Comment by QuantAkt** [comment]
Looks great, but this does not seem to work, if you would like the listing of a Python script? I tried the above commands, simply replacing where applicable ".Rmd" with the relevant ".py" but neither approach produces a listing. Just empty space. Shouldn't this be literal quoting, or is there some interpretation of the code involved?

**Reply by yihui** [reply]
I'm not sure what you tried exactly. It will be easier for me to help if you could provide a minimal reproducible example. Here is mine:

````md
---
title: Python code in appendix
---

```{r, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

```{python, chunk-a}
1 + 1
```

```{python, chunk-b}
2 + 2
```

```{python, ref.label=knitr::all_labels(engine == 'python'), echo = T, eval = F}
```
````

**Reply by QuantAkt** [reply]
Of course, you are right. Here are my attempts. I did not include the additional file "boring_function.py" but its contents should be clear?

````
---
title: "inclusiveexample"
---
# define stuff
```{r}
library(reticulate)
```

```{r}
source_python("boring_function.py")

# Boring function is defined as
# def boring_function(x):
#    return(x + 1)

```

# do stuff
```{python}
tmp = boring_function(1)
print(tmp)
```

# Appendix

For completeness' sake all boring functions should be listed here.
Ideally the listing should be formatted exactly like this "native" definition:
```{python fundef}
def boring_function(x):
    return(x + 1)
```

Does nothing
```{python code = readLines(knitr::purl("boring_function.py", documentation = 1)), echo = T, eval = F}

```

Also does nothing
```{python, child="boring_function.py", eval=F}

```

Does something but not the right thing respectively not for the right reasons
```{python, ref.label=knitr::all_labels(engine == 'python'), echo = T, eval = F}

```

This is the best I could manage this far. But the formatting is off.
```{r, echo=FALSE, results='asis'}
res <- knitr::knit_child('boring_function.py', quiet = TRUE)
cat(res, sep = '\n')
```

````

**Reply by QuantAkt** [reply]
Just to be on the safe side: 

I want to include the listing of an external python file within an Rmarkdown document. This listing should be formatted in the same way as a "native" definition, i.e. a definition made within the Rmarkdown doc, not in an external file which is sourced.

**Reply by yihui** [reply]
I think this should be what you were looking for:

````md
First, list the python file:

```{python, file='boring_function.py', echo=T, eval=F}
```

Then list the rest of python code chunks:

```{python, ref.label=knitr::all_labels(engine == 'python'), echo=T, eval=F}
```
````

**Reply by QuantAkt** [reply]
That easy! Great!!
